src/libAtomVM/bitstring.c

```diff
@@ -119,10 +140,73 @@ bool bitstring_utf8_encode(avm_int_t c, uint8_t *buf, size_t *out_size)
     return true;
 }
 
+bool bitstring_utf8_decode(int32_t *c, uint8_t *buf, size_t len, size_t *out_size)
+{
+    int32_t v = 0;
+    if (len == 0) {
+        return false;
+    } else if (len >= 4 && (*buf & 0xF8) == 0xF0 && ((*(buf + 1) & 0xC0) == 0x80) && ((*(buf + 2) & 0xC0) == 0x80) && ((*(buf + 3) & 0xC0) == 0x80)) {
+        v |= (*buf & 0x07) << 18;
+        v |= (*(buf + 1) & 0x3F) << 12;
+        v |= (*(buf + 2) & 0x3F) << 6;
+        v |= (*(buf + 3) & 0x3F);
+        if (invalid_codepoint(v)) {
```

Comment:
This test could be limited to ˋv > 0x10FFFF` and combined with the next one.

Reply:
Are you sure about that?  Isn't it still possible to encode a surrogate, a negative value, and a value greater than 0x10FFFF (1114111) in 21 bits (2^21 = 2097152)?
```
// | U+10000  | 11110xxx | 10xxxxxx | 10xxxxxx | 10xxxxxx |
```

Reply:
(or is the point that if it were a surrogate it would be overlong, by definition?)